lib/luacrypto_rng.c

```diff
@@ -0,0 +1,207 @@
+/*
+* SPDX-FileCopyrightText: (c) 2025 jperon <cataclop@hotmail.com>
+* SPDX-License-Identifier: MIT OR GPL-2.0-only
+*/
+
+/***
+* Low-level Lua interface to the Linux Kernel Crypto API for synchronous
+* Random Number Generators (RNG).
+*
+* This module provides a `new` function to create RNG objects,
+* which can then be used for generating random bytes.
+*
+* @module crypto_rng
+*/
+
+#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
+
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/crypto.h>
+#include <crypto/rng.h>
+#include <linux/err.h>
+#include <linux/limits.h>
+#include <linux/slab.h>
+
+#include <lua.h>
+#include <lualib.h>
+#include <lauxlib.h>
+#include <lunatik.h>
+
+#include "luacrypto.h"
+
+LUNATIK_PRIVATECHECKER(luacrypto_rng_check, struct crypto_rng *);
+
+LUACRYPTO_RELEASER(rng, struct crypto_rng, crypto_free_rng, NULL);
+
+/***
+* RNG object methods.
+* These methods are available on RNG objects created by `crypto_rng.new()`.
+* @see crypto_rng.new
+* @type crypto_rng
+*/
+
+/***
+* Generates a specified number of random bytes, with optional seed.
+* @function crypto_rng:generate
+* @tparam integer num_bytes The number of random bytes to generate.
+* @tparam[opt] string seed Optional seed material to mix into the RNG. If nil or omitted, no explicit seed is used.
+* @treturn string A binary string containing the generated random bytes.
+* @raise Error on failure (e.g., allocation error, crypto API error).
+*/
+static int luacrypto_rng_generate(lua_State *L)
+{
+	struct crypto_rng *tfm = luacrypto_rng_check(L, 1);
+	size_t num_bytes = (size_t)luaL_checkinteger(L, 2);
+	luaL_argcheck(L, num_bytes >= 0, 2, "number of bytes must be non-negative");
+
+	size_t seed_len = 0;
+	const char *seed_data = lua_tolstring(L, 3, &seed_len);
+
+	luaL_Buffer b;
+	char *buffer = luaL_buffinitsize(L, &b, num_bytes);
+
+	lunatik_try(L, crypto_rng_generate, tfm, seed_data, (unsigned int)seed_len, (u8 *)buffer, (unsigned int)num_bytes);
+
+	luaL_pushresultsize(&b, num_bytes);
+	return 1;
+}
+
+/***
+* Generates a specified number of random bytes.
+* This function does not take an explicit seed as an argument.
+* @function crypto_rng:getbytes
+* @tparam integer num_bytes The number of random bytes to generate.
+* @treturn string A binary string containing the generated random bytes.
+* @raise Error on failure (e.g., allocation error, crypto API error).
+*/
+static int luacrypto_rng_getbytes(lua_State *L)
+{
+	struct crypto_rng *tfm = luacrypto_rng_check(L, 1);
+	unsigned int num_bytes = (unsigned int)luaL_checkinteger(L, 2);
+
+	luaL_argcheck(L, num_bytes >= 0 && num_bytes <= UINT_MAX, 2, "out of bounds");
+
+	if (num_bytes == 0) {
+		lua_pushlstring(L, "", 0);
+		return 1;
+	}
```

Comment:
this is the very same case.. btw, what happens when you pass zero here and above? does it fail safely?